components/browser/domains/src/main/java/mozilla/components/browser/domains/autocomplete/Providers.kt

```diff
@@ -23,40 +23,40 @@ enum class DomainList(val listName: String) {
 /**
  * Provides autocomplete functionality for domains based on provided list of assets (see [Domains]).
  */
-class ShippedDomainsProvider : BaseDomainAutocompleteProvider(DomainList.DEFAULT) {
-    override fun initialize(context: Context) {
-        scope.launch {
-            domains = async { Domains.load(context).into() }.await()
-        }
-    }
-}
+class ShippedDomainsProvider : BaseDomainAutocompleteProvider(DomainList.DEFAULT, Domains.asLoader())
 
 /**
  * Provides autocomplete functionality for domains based on a list managed by [CustomDomains].
  */
-class CustomDomainsProvider : BaseDomainAutocompleteProvider(DomainList.CUSTOM) {
-    override fun initialize(context: Context) {
-        scope.launch {
-            domains = async { CustomDomains.load(context).into() }.await()
-        }
-    }
-}
+class CustomDomainsProvider : BaseDomainAutocompleteProvider(DomainList.CUSTOM, CustomDomains.asLoader())
 
 interface DomainAutocompleteProvider {
     fun getAutocompleteSuggestion(query: String): DomainAutocompleteResult?
 }
 
+typealias DomainsLoader = (Context) -> List<Domain>
+
+private fun Domains.asLoader(): DomainsLoader = { context: Context -> load(context).into() }
+private fun CustomDomains.asLoader(): DomainsLoader = { context: Context -> load(context).into() }
+
 /**
  * Provides common autocomplete functionality powered by domain lists.
+ *
+ * @param list source of domains
+ * @param domainsLoader provider for all available domains
  */
-abstract class BaseDomainAutocompleteProvider(private val list: DomainList) :
-    DomainAutocompleteProvider {
-    internal val scope = CoroutineScope(Dispatchers.IO)
-    // We compute 'domains' on the worker thread; make sure it's immediately visible on the UI thread.
-    @Volatile
+open class BaseDomainAutocompleteProvider(
+    private val list: DomainList,
+    private val domainsLoader: DomainsLoader
+) : DomainAutocompleteProvider, CoroutineScope by CoroutineScope(Dispatchers.IO) {
+
     var domains: List<Domain> = emptyList()
```

Comment:
Did you intentionally remove the `@Volatile` annotation here? With reading/writing the list from different threads it looks like we still need it?

Reply:
Yes, I've removed it intentionally but I was wrong doing this. :)
Will return it back.